### PR TITLE
fix(middleware/cors): Handling and wildcard subdomain matching

### PR DESCRIPTION
**Pull request by sixcolors**
# PR Description

## Summary of Changes
This pull request enhances the CORS middleware in the Fiber web framework by introducing improved origin matching, subdomain support, and comprehensive documentation updates.

## Changes Made
1. **Origin Matching Improvements:**
   - Refactored the `utils.validateDomain` function to handle wildcard subdomain patterns more accurately.
   - Added tests to ensure correct validation of various origin patterns.

2. **Subdomain Support:**
   - Modified the CORS middleware to support subdomain matching for the `AllowOrigins` configuration.
   - Added tests to verify the behavior of subdomain matching.

3. **Documentation Updates:**
   - Added detailed documentation explaining how the CORS middleware works, including the handling of different `AllowOrigins` configurations.
   - Highlighted security considerations and best practices to avoid potential vulnerabilities.
   - Included information on handling misconfigurations related to wildcard origins and credentials.
   - Updated the documentation to reflect changes in the middleware's behavior.

4. **Improved CORS Handling:**
   - Introduced the `setCORSHeaders` function to streamline the management of common behaviors in both simple and pre-flight requests.
   - Restructured the handler function to consistently apply CORS rules through the newly added setCORSHeaders function.
## How It Works Section
- Added a new section in the documentation explaining the functionality of the CORS middleware.
- Described the process of adding CORS headers to responses based on the configuration.
- Clarified the handling of preflight requests and the dynamic evaluation of allowed origins.

## Security Considerations Section
- Introduced a section emphasizing security considerations when configuring CORS.
- Highlighted potential risks of allowing all origins, credentials exposure, and exposing sensitive headers.
- Provided guidance on proper configurations to mitigate security threats.

## Testing Updates
- Expanded test coverage to include scenarios related to subdomains, origin validation, and wildcard patterns.
- Added tests for improved coverage of the CORS middleware.

## Notes for Reviewers
- Please review the changes made to the `utils.validateDomain` function and ensure accuracy in handling wildcard subdomain patterns.
- Verify the correctness of the subdomain matching behavior in the CORS middleware.
- Verify the correctness of the middleware behavior when making simple requests.
- Thoroughly review the documentation updates for clarity, completeness, and adherence to best practices.
- This is targeting v2 branch, we will also need to cherry pick and merge into v3/main.

This pull request aims to enhance the functionality, security, and documentation of the CORS middleware in Fiber. Your feedback and suggestions are highly appreciated.

# Commits

fix(middleware/cors): handling and wildcard subdomain matching
docs(middleware/cors): add 'How it works' and 'Security Considerations'
chore: grammar
chore: fix misspelling
test(middleware/cors): combine Invalid_Origins tests
refactor(middleware/cors): headers handling

## Summary by CodeRabbit

- **New Features**
	- Enhanced CORS middleware support including subdomain matching and dynamic origin evaluation.
- **Bug Fixes**
	- Corrected `Access-Control-Allow-Origin` header handling based on `AllowCredentials`.
- **Refactor**
	- Improved CORS middleware code organization with a new function for setting headers.
- **Tests**
	- Added and adjusted tests for various CORS configurations and scenarios.
- **Documentation**
	- Updated CORS middleware documentation with security considerations and configuration guidance.

### CODE REVIEW

**Comment by sixcolors** [comment]
For reference when reviewing: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

File: docs/api/middleware/cors.md [hunk 1]
```diff
@@ -73,7 +73,7 @@
...
-| AllowOrigins     | `string`                   | AllowOrigin defines a comma separated list of origins that may access the resource.                                                                                                                                                                                                                                                                                | `"*"`                              |
+| AllowOrigins     | `string`                   | AllowOrigins defines a comma separated list of origins that may access the resource. This supports subdomain matching, so you can use a value like "https://.example.com" to allow any subdomain of example.com to submit requests.                                                                                                                                | `"*"`                              |
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:76]: right it supports sub.sub.example.com and so on.

File: docs/api/middleware/cors.md [hunk 2]
```diff
@@ -101,3 +101,51 @@
...
+:::note
+In our CORS middleware, we specifically prevent `Access-Control-Allow-Credentials` from being `true` when `Access-Control-Allow-Origin` is set to the wildcard (`*`). This prevents potential security risks associated with allowing credentials to be shared with all origins. 
+
+When using `AllowOrigins`, a configuration check will cause a panic if `Access-Control-Allow-Credentials` is `true` and `Access-Control-Allow-Origin` is set to the wildcard.
+:::
+
+:::caution
+Be extra careful when using `AllowOriginsFunc`. Make sure to properly validate the origin to prevent potential security risks.
+
+When using `AllowOriginsFunc`, the `Access-Control-Allow-Origin` header will always be set to the origin header if the func returns `true`, which can bypass such protections if you return `true` in all situations.
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:150]: Please make a suggestion.

File: docs/api/middleware/cors.md [hunk 3]
```diff
@@ -8,7 +8,7 @@ CORS middleware for [Fiber](https://github.com/gofiber/fiber) that can be used t
 
 The middleware conforms to the `access-control-allow-origin` specification by parsing `AllowOrigins`. First, the middleware checks if there is a matching allowed origin for the requesting 'origin' header. If there is a match, it returns exactly one matching domain from the list of allowed origins.
 
-For more control, `AllowOriginsFunc` can be used to programatically determine if an origin is allowed. If no match was found in `AllowOrigins` and if `AllowOriginsFunc` returns true then the 'access-control-allow-origin' response header is set to the 'origin' request header.
+For more control, `AllowOriginsFunc` can be used to programmatically determine if an origin is allowed. If no match was found in `AllowOrigins` and if `AllowOriginsFunc` returns true then the 'access-control-allow-origin' response header is set to the 'origin' request header.
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:11]: Please make a suggestion.

File: docs/api/middleware/cors.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:76]: Please make a suggestion.

**Comment by ReneWerner87** [comment]
image


@sixcolors you can look for the last adjustments, so far it actually looks quite good

**Comment by ReneWerner87** [comment]
maybe you can also take another look at why it has become so much slower

**Comment by sixcolors** [comment]
> maybe you can also take another look at why it has become so much slower

Will take a look.

**Comment by sixcolors** [comment]
@ReneWerner87 performance:

PR:

```bash
(fix-cors-allow-origins) » go test -v -run=^$ -bench=Benchmark_CORS -benchmem -count=4
goos: darwin
goarch: arm64
pkg: github.com/gofiber/fiber/v2/middleware/cors
Benchmark_CORS_NewHandler
Benchmark_CORS_NewHandler-12                                     4294808               278.0 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandler-12                                     4296440               278.3 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandler-12                                     4249886               281.4 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandler-12                                     4271992               281.4 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerParallel
Benchmark_CORS_NewHandlerParallel-12                            31935631                34.49 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            33806983                34.10 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            33057889                40.40 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            29357768                34.79 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin
Benchmark_CORS_NewHandlerSingleOrigin-12                         4080190               290.2 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         4133065               292.4 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         4110668               290.3 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         4096338               287.7 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel
Benchmark_CORS_NewHandlerSingleOriginParallel-12                37063442                31.96 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                34628786                33.69 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                35921152                32.35 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                36194454                38.67 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcard
Benchmark_CORS_NewHandlerWildcard-12                             4668529               255.8 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             4722679               255.7 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             4677729               253.6 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             4746825               254.1 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel
Benchmark_CORS_NewHandlerWildcardParallel-12                    42866646                29.58 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    40805569                28.18 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    42172837                27.54 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    43017433                28.31 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflight
Benchmark_CORS_NewHandlerPreflight-12                            2801908               426.3 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2827065               424.1 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2813889               424.5 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2821395               424.7 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel
Benchmark_CORS_NewHandlerPreflightParallel-12                   23274964                51.78 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   23190030                51.70 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   23248810                51.74 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   22682397                51.87 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2819092               426.1 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2807754               424.9 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2817894               424.0 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2825562               423.7 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       24872613                48.45 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       23707059                48.23 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       24249118                48.36 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       24710340                48.11 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard
Benchmark_CORS_NewHandlerPreflightWildcard-12                    3104851               384.0 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    3112880               383.1 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    3122064               384.1 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    3122048               383.0 ns/op             0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           28592589                41.11 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           28256901                41.37 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           28957732                41.74 ns/op            0 B/op          0 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           28696690                41.29 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/gofiber/fiber/v2/middleware/cors     66.870s
```

Main:

```bash
(main) » go test -v -run=^$ -bench=Benchmark_CORS -benchmem -count=4     sixcolors@Jasons-MacBook-Pro-2023
goos: darwin
goarch: arm64
pkg: github.com/gofiber/fiber/v3/middleware/cors
Benchmark_CORS_NewHandler
Benchmark_CORS_NewHandler-12                                     5094988               217.4 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandler-12                                     5520282               216.8 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandler-12                                     5507739               219.0 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandler-12                                     5412978               220.0 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerParallel
Benchmark_CORS_NewHandlerParallel-12                            31298053                40.20 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            32613938                42.57 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            29129778                41.67 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerParallel-12                            34623457                39.94 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin
Benchmark_CORS_NewHandlerSingleOrigin-12                         5281987               227.5 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         5159940               229.1 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         5290394               226.0 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOrigin-12                         5332168               226.8 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel
Benchmark_CORS_NewHandlerSingleOriginParallel-12                31422558                41.33 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                31525850                42.30 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                29251496                42.56 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerSingleOriginParallel-12                31469149                41.65 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcard
Benchmark_CORS_NewHandlerWildcard-12                             6547953               179.9 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             6564964               180.2 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             6612141               180.4 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcard-12                             6624924               179.5 ns/op            16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel
Benchmark_CORS_NewHandlerWildcardParallel-12                    39140804                33.59 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    38824480                34.73 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    38681133                38.34 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerWildcardParallel-12                    32569826                39.27 ns/op           16 B/op          1 allocs/op
Benchmark_CORS_NewHandlerPreflight
Benchmark_CORS_NewHandlerPreflight-12                            1973256               593.3 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2072100               577.6 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2075713               586.1 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflight-12                            2029602               582.7 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel
Benchmark_CORS_NewHandlerPreflightParallel-12                   13736329                94.63 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   12909877                95.58 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   11570576               106.2 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightParallel-12                   12331178               104.4 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2147710               559.1 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2139936               553.9 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2166442               553.0 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOrigin-12                2162149               554.5 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       13443846                93.33 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       13502077                92.76 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       14054209               100.6 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightSingleOriginParallel-12       11905135               100.8 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard
Benchmark_CORS_NewHandlerPreflightWildcard-12                    2291680               527.4 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    2335252               516.6 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    2332082               516.8 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcard-12                    2312286               520.9 ns/op            48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           12608054                84.95 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           14658009                84.15 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           14614734                83.24 ns/op           48 B/op          3 allocs/op
Benchmark_CORS_NewHandlerPreflightWildcardParallel-12           14585637                80.84 ns/op           48 B/op          3 allocs/op
PASS
ok      github.com/gofiber/fiber/v3/middleware/cors     70.468s
```

**Comment by sixcolors** [comment]
@gaby the question does remain about the `https://.example.com` subdomain matching (Django Style) vs `https://*.example.com` our old style (and rs/cors style) which we did not have in the docs.

Or the Spring style

> Wildcards are not authorized in allowOrigins, but alternatively the allowOriginPatterns property may be used to match to a dynamic set of origins.

**Comment by ReneWerner87** [comment]
@sixcolors could you use the utils string functions instead of the std string pkg in v2
image

TrimLeft, TrimRight, ToLower ...

**Review by ReneWerner87 — APPROVED** [review]
LGTM

**Comment by ReneWerner87** [comment]
> @gaby the question does remain about the `https://.example.com` subdomain matching (Django Style) vs `https://*.example.com` our old style (and rs/cors style) which we did not have in the docs.
> 
> Or the Spring style
> 
> > Wildcards are not authorized in allowOrigins, but alternatively the allowOriginPatterns property may be used to match to a dynamic set of origins.

i like the old style -> in my opinion what i have seen in most frameworks

**Comment by sixcolors** [comment]
> > @gaby the question does remain about the `https://.example.com` subdomain matching (Django Style) vs `https://*.example.com` our old style (and rs/cors style) which we did not have in the docs.
> > Or the Spring style
> > > Wildcards are not authorized in allowOrigins, but alternatively the allowOriginPatterns property may be used to match to a dynamic set of origins.
> 
> i like the old style -> in my opinion what i have seen in most frameworks

I’ll modify to the old style but with * only being a wildcard for a subdomain eg `https://*.example..com` and not `http://*example.com`, `https://*`, or `https://www.*.com` etc. If someone needs that they would need to use allowOriginsFunc

File: docs/api/middleware/cors.md [hunk 4]
```diff
@@ -72,8 +70,8 @@
...
+| AllowOriginsFunc | `func(origin string) bool` | AllowOriginsFunc defines a function that will set the 'Access-Control-Allow-Origin' response header to the 'origin' request header when returned true. This allows for dynamic evaluation of allowed origins. Note if AllowCredentials is true, wildcard origins will be not have the 'Access-Control-Allow-Credentials' header set to 'true'.                     | `nil`                              |
+| AllowOrigins     | `string`                   | AllowOrigins defines a comma separated list of origins that may access the resource. This supports subdomain matching, so you can use a value like "https://*.example.com" to allow any subdomain of example.com to submit requests.                                                                                                                               | `"*"`                              |
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:74]: it's covered in more detail elsewhere in the doc, this is just a table of config options.

File: docs/api/middleware/cors.md [hunk 5]
```diff
@@ -6,11 +6,9 @@ id: cors
 
 CORS middleware for [Fiber](https://github.com/gofiber/fiber) that can be used to enable [Cross-Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) with various options.
 
-The middleware conforms to the `access-control-allow-origin` specification by parsing `AllowOrigins`. First, the middleware checks if there is a matching allowed origin for the requesting 'origin' header. If there is a match, it returns exactly one matching domain from the list of allowed origins.
+The middleware conforms to the `Access-Control-Allow-Origin` specification by parsing `AllowOrigins`. First, the middleware checks if there is a matching allowed origin for the requesting 'origin' header. If there is a match, it returns exactly one matching domain from the list of allowed origins. If there is no match, the middleware does not add the `Access-Control-Allow-Origin` header to the response.
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:9]: sure suggest one.

File: docs/api/middleware/cors.md [hunk 6]
```diff
@@ -72,8 +70,8 @@
...
-| AllowOrigins     | `string`                   | AllowOrigin defines a comma separated list of origins that may access the resource.                                                                                                                                                                                                                                                                                | `"*"`                              |
+| AllowOriginsFunc | `func(origin string) bool` | AllowOriginsFunc defines a function that will set the 'Access-Control-Allow-Origin' response header to the 'origin' request header when returned true. This allows for dynamic evaluation of allowed origins. Note if AllowCredentials is true, wildcard origins will be not have the 'Access-Control-Allow-Credentials' header set to 'true'.                     | `nil`                              |
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:73]: suggest fix

File: docs/api/middleware/cors.md [hunk 7]
```diff
@@ -101,3 +83,71 @@
...
 }
 ```
+
+## Subdomain Matching  
+
+The `AllowOrigins` configuration supports matching subdomains at any level. This means you can use a value like `"https://*.example.com"` to allow any subdomain of `example.com` to submit requests, including multiple subdomain levels such as `"https://sub.sub.example.com"`.
+
+### Example  
+
+If you want to allow CORS requests from any subdomain of `example.com`, including nested subdomains, you can configure the `AllowOrigins` like so:
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:95]: The previous example was highly insecure due to it resulting in a wildcard match, which bypassed the prohibition on AllowCredentials with wildcard origins.

Comment (sixcolors) [reply docs/api/middleware/cors.md:95]: The example I removed was the AllowOriginsFunc example

[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md]: The example I removed would create an effective wildcard match, which would bypass the prohibition on wildcard origins with AllowCredentials = true.

 ```go
 app.Use(cors.New())

 app.Use(cors.New(cors.Config{
     AllowOriginsFunc: func(origin string) bool {
         return os.Getenv("ENVIRONMENT") == "development"
     },
 }))
 ```
 
In my opinion, the AllowOriginsFunc is a dangerous feature, and only individuals with specific needs and a thorough understanding of its implications should use it. The security considerations outlined in the documentation, along with the notes in the configuration options, should be enough to caution users about the potential dangers. Including a basic example of this feature, especially an insecure one, will cause more harm than good.

Using a database to look up the origins is a more secure example and a valid use case for this feature. However, including too much information in the documentation can be overwhelming for junior developers and cause them to miss more important notes.

[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md]: Sure, suggest an example for using the AllowOriginsFunc using a DB to match AllowedOrigins, make sure it's secure and not overwhelming.

**Comment by sixcolors** [comment]
@jub0bs I would appreciate your review as I had gone back to see why some decisions about this middleware were made and https://github.com/gofiber/fiber/issues/2390 explained some decisions I was questioning. 

I think this PR improves the situation while being backward compatible within the documented api. I do plan a complete rewrite for v3. 

Thanks!

**Review by gaby — APPROVED** [review]
LGTM

**Comment by gaby** [comment]
@sixcolors Should we add the changes from #2908 in this PR?

I'm going to update that PR to rename the fields first.

**Comment by ReneWerner87** [comment]
> I'm going to update that PR to rename the fields first.

no new features for v2 , only bugfixes and security-related changes

**Pull request merged**